File: C-algorithm/ProgrammingPearls/sort/quickSort.c
/**
	快速排序
*/

void qsort(int *arr, int lo, int hi){
	if (lo >= hi){
		return ;	// 一个元素
	}
	
	int pivot = partition(arr, lo, hi);
	
	qsort(arr, lo, pivot - 1);
	qsort(arr, pivot + 1, hi);
}

/*****************************************************/
void qsort(int *arr, int lo, int hi){
	if (lo >= hi){
		return ;	// 一个元素
	}
	
	int mid = lo;
	for (int i = l + 1; i <= hi; i++){
		if (arr[i] < arr[lo]){
			swap(arr[i], arr[++mid]);	// mid != i 才需要交换，否则自身交换。
		}
	}
	swap(arr[lo], arr[mid]);
	
	qsort(arr, lo, pivot - 1);
	qsort(arr, pivot + 1, hi);
}

File: java/crackingTheCodingInterview/sort/quicksort.java
/**
	快速排序
*/
void quicksort(int[] array, int left, int right){
	int index = partition(array, left, right);
	if (left < index - 1){
		quicksort(arr, left, index - 1);
	}
	if (index < right - 1){
		quicksort(array, index + 1, right);
	}
}

int partition(int[] array, int left, int right){
	int pivot = array[(left + right) / 2];
	while (left <= right){
		while (array[left] <= pivot){
			left++;
		}
		while (array[right] >= pivot){{
			right--;
		}
		if (left <= right){
			swap(array, left, right);
			left++;
			right--;
		}
	}
	return left;
}